Update the module guide - Closes #1076 (#1102)
LiskHQ/lisk-docs: modules/ROOT/pages/guides/app-development/module.adoc

:url_references_schemas: references/schemas.adoc
:url_tutorials_hello: tutorials/hello-world.adoc

How to create a new module for a blockchain application.

.Prerequisites
[NOTE]
====
To follow this guide, the following criteria is assumed:

* Lisk Commander is installed, and a basic blockchain application is already initialized, as explained in the previous guide xref:{url_guides_setup}[].
====

== Generating the module skeleton

Use Lisk Commander to generate the skeleton for the new module:

[source,bash]
----
$ lisk generate:module --help
Creates a module skeleton for the given name and id.

USAGE
  $ lisk generate:module MODULENAME MODULEID

ARGUMENTS
  MODULENAME  Module name.

>   MODULEID    Module Id, should be atleast 1000 or more

Reviewer: ```suggestion
  MODULEID    Module Id, should be at least 1000 or more
```

OPTIONS
  -t, --template=template  Template to bootstrap the application. It will read from `.liskrc.json` or use `lisk-ts` if not found.

EXAMPLE
  generate:module nft 5000
----

The command `lisk generate:module` expects 2 arguments:

. *Module name*: The name of the module.
The name is used for example to subscribe to events, or to invoke actions of the module.

> Needs to be a string that only consists of small and capital letters [a-z,A-Z].

Reviewer: ```suggestion
Needs to be a string that only consists of upper and lower case letters [a-z, A-Z].
```

> No numbers, hyphen, etc allowed.

Reviewer: ```suggestion
No numbers, hyphens, etc., are allowed.
```

. *Module ID*: The ID of the module.
Needs to be unique within the application.
Minimum value is `1000`.

> As example, we use the module name `nft` and the module ID `1000`:

Reviewer: ```suggestion
As an example, we use the module name `nft` and the module ID `1000`:
```

[[generate-module]]
[source,bash]
----
lisk generate:module nft 1000
----

This will generate the following files:

----
Creating module skeleton with module name "nft" and module ID "1000"
Using template "lisk-ts"
Generating module skeleton.
Registering module...
identical .liskrc.json
   create src/app/modules/nft/nft_module.ts
   create test/unit/modules/nft/nft_module.spec.ts

No change to package.json was detected. No package manager install will be executed.

Your module is created and ready to use.
----

It will also automatically register the module with the application, by adding it to `src/app/modulees.ts`.

.src/app/modules.ts
[source,typescript]
----
import { Application } from 'lisk-sdk';
import { NftModule } from "./modules/nft/nft_module";

export const registerModules = (app: Application): void => {
    app.registerModule(NftModule);
};

----

The file `nft_module.ts` contains the module skeleton and the file `nft_module.spec.ts` contains the related unit test for the new module.

> First take a look at the module skeleton in `nft_module.ts`:

Reviewer: ```suggestion
The module skeleton can be viewed in `nft_module.ts`:
```

    }

    public async afterBlockApply(_input: AfterBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async beforeTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext) {
        // Get any data from genesis block, for example get all genesis accounts
        // const genesisAccoounts = genesisBlock.header.asset.accounts;
    }
}
----

> The command `generate:module` already created the class `NftModule` which contains already skeletons for the most important components of a module.

Reviewer 2: ```suggestion
The command `generate:module` already created the class `NftModule` which contains skeletons for the most important components of a module.
```

> The only properties which are set at the moment are the module ID and the module name, which we defined before.

Reviewer: ```suggestion
The only properties which are set at this point are the module ID and the module name, which were defined previously.
```

> In fact, these 2 properties make it already a complete module, which can be registered with the application.

Reviewer: ```suggestion
In fact, with these 2 properties, it is already a complete module that can be registered with the application.
```

> The only problem is, that this module isn't doing anything yet.

Reviewer: ```suggestion
However, this module is not performing any functions yet.
```

> To give the module a purpose, it is needed to implement certain logic inside of the module.

Reviewer: ```suggestion
To give the module a purpose, it is necessary to implement certain logic inside of the module.
```

----

[source,typescript]
----
public actions = {
    // Example below
    // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
    // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
};
----

== Reducers

A list of actions that other modules of the application can invoke.

NOTE: Reducers can only be invoked by other modules, *not* by plugins.

[source,typescript]
----
public reducers = {
    // Example below
    // getBalance: async (
    // 	params: Record<string, unknown>,
    // 	stateStore: StateStore,
    // ): Promise<bigint> => {
    // 	const { address } = params;
    // 	if (!Buffer.isBuffer(address)) {
    // 		throw new Error('Address must be a buffer');
    // 	}
    // 	const account = await stateStore.account.getOrDefault<TokenAccount>(address);
    // 	return account.token.balance;
    // },
};
----

== Events

A list of events that this module is able to emit.

> Modules, plugins and external services can subscribe to these events.

Reviewer: ```suggestion
Modules, plugins, and external services can subscribe to these events.
```

TIP: See the xref:{url_rpc_endpoints}[] page for more information.

[source,typescript]
----
public events = [
    // Example below
    // 'nft:newBlock',
];
----

[TIP]

====

> What events, actions and reducers are used within a module, or if these interfaces are actually required at all, will be a specific individual requirement for every module; as it is heavily dependant on which functionality the module intends to provide to the application.

Reviewer: ```suggestion
Which events, actions, and reducers are used within a module, or if these interfaces are actually required at all, will be a specific individual requirement for every module; as it is heavily dependant on which functionality the module intends to provide to the application.
```

> The best way to understand the necessary requirements here is to look at existing examples in the Lisk SDK default modules, or examples of other blockchain applications built with the Lisk SDK, for example the xref:{url_tutorials_hello}[] application.

Reviewer: ```suggestion
The easiest way to understand the necessary requirements here is to look at existing examples in the Lisk SDK default modules, or examples of other blockchain applications built with the Lisk SDK, for example, the xref:{url_tutorials_hello}[] application.
```

}
----

== Defining an account schema

> In some cases, the new module will require to store some new data in the user accounts.

Reviewer: ```suggestion
In some cases, the new module will require storing some new data in the user accounts.
```

If that is the case, it is required to define the corresponding account schema in the module.

The account schema defines which properties are added to user accounts by the module.

TIP: For more information about schemas and how they are used in the Lisk SDK, check out the xref:{url_references_schemas}[].

The module-specific properties will be added under a key that is named after the module name.

> So e.g. in case the module name is `nft`, the data will be stored under the key "nft" in each user account.

Reviewer: ```suggestion
So for example, in case the module name is `nft`, the data will be stored under the key "nft" in each user account.
```

[IMPORTANT]

====

> If a module includes an account schema, it is needed to update the genesis block after registering the module with the application.

Reviewer: ```suggestion
If a module includes an account schema, it is necessary to update the genesis block after registering the module with the application.
```

> You can use the application CLI to generate a new genesis block like this:

Reviewer: ```suggestion
The application CLI can be used to generate a new genesis block as shown below:
```
